app/src/main/java/me/hufman/androidautoidrive/music/controllers/SpotifyAppController.kt

```diff
@@ -242,17 +249,115 @@ class SpotifyAppController(context: Context, val remote: SpotifyAppRemote, val w
 	 * not authorized then the the [QueueMetadata] is created from the app remote API.
 	 */
 	fun createLikedSongsQueueMetadata() {
+		if (createQueueMetadataJob?.isActive == true) {
+			createQueueMetadataJob?.cancel()
+		}
+
 		createQueueMetadataJob = GlobalScope.launch(defaultDispatcher) {
 			queueItems = webApi.getLikedSongs(this@SpotifyAppController) ?: emptyList()
+
 			if (queueItems.isNotEmpty()) {
-				queueMetadata = QueueMetadata("Liked Songs", null, queueItems)
-				loadQueueCoverart()
+				queueMetadata = QueueMetadata(queueTitle, null, queueItems)
+
+				val likedSongsTemporaryPlaylistStateKey = AppSettings.KEYS.SPOTIFY_LIKED_SONGS_PLAYLIST_STATE
+				val likedSongsStateJson = appSettings[likedSongsTemporaryPlaylistStateKey]
+
+				val temporaryPlaylistState = if (likedSongsStateJson.isBlank()) {
+					Log.d(TAG, "No previous liked songs state found.")
+					createTemporaryPlaylistState(SpotifyWebApi.LIKED_SONGS_PLAYLIST_NAME)
+				} else {
+					Log.d(TAG, "Found previous liked songs state.")
+					loadTemporaryPlaylistState(likedSongsStateJson)
+				}
+				if (temporaryPlaylistState == null) {
+					Log.e(TAG, "Error creating liked songs playlist, falling back to app remote API")
+					queueItems = emptyList()
+					createQueueMetadata(PlayerContext(queueUri, queueTitle, null, null))
+					return@launch
+				}
+
+				saveTemporaryPlaylistState(temporaryPlaylistState, likedSongsTemporaryPlaylistStateKey)
 
 				callback?.invoke(this@SpotifyAppController)
 			} else {
-				createQueueMetadata(PlayerContext(queueUri, "Liked Songs", null, null))
+				createQueueMetadata(PlayerContext(queueUri, queueTitle, null, null))
+			}
+		}
+	}
+
+	/**
+	 * Creates a [TemporaryPlaylistState] for the supplied playlist name, creating the playlist if it
+	 * is not present and adding the queueItems in context to it. If the playlist creation fails
+	 * then null is returned.
+	 */
+	private suspend fun createTemporaryPlaylistState(playlistName: String): TemporaryPlaylistState? {
+		val queueItemsHashCode = queueItems.hashCode().toString()
+		val existingPlaylistUri = webApi.getPlaylistUri(playlistName)
+		val playlistUri: String
+		val playlistId: String
+		if (existingPlaylistUri == null) {
+			Log.d(TAG, "No user playlist for $playlistName found, creating a new one.")
+			val uri = webApi.createPlaylist(playlistName, L.MUSIC_TEMPORARY_PLAYLIST_DESCRIPTION) ?: return null
+			webApi.addSongsToPlaylist(uri.id, queueItems)
+			playlistUri = uri.uri
+			playlistId = uri.id
+		} else {
+			Log.d(TAG, "User playlist for $playlistName found, using existing playlist.")
+			playlistUri = existingPlaylistUri.uri
+			playlistId = existingPlaylistUri.id
+		}
+
+		val temporaryPlaylistState = TemporaryPlaylistState(queueItemsHashCode, playlistUri, playlistId, queueTitle)
+
+		queueUri = temporaryPlaylistState.playlistUri
+
+		val recentlyPlayedUri = "com.spotify.recently-played"
```

Comment:
This duplicated piece of code could be combined with the regular `loadQueueCoverart` if the `webApi.setPlaylistImage` was conditional (perhaps `playerContext.title != SpotifyWebApi.LIKED_SONGS_PLAYLIST_NAME`).
Alternatively, `loadQueueCoverart` could be converted to a suspend function with a CompletableDeferred, similar to how the [adapter to the native browse functionality works](https://github.com/hufman/AndroidAutoIdrive/blob/e5968c276b587537e12f791444386fd82ff3dbbf/app/src/main/java/me/hufman/androidautoidrive/music/MusicBrowser.kt#L123), and return the coverart instead of replacing the queueMetadata.

Reply:
Ah I like the concept of the loadQueueCoverart just returning a Bitmap as a suspend function with a CompletableDeferred, that makes it more generic and allows the consumers to define what to do with that coverArt image which is better in my opinion than trying to cram too much logic inside for all the different scenarios. I updated the function and renamed it now as ```getQueueCoverArt()``` and updated the logic throughout that call this method